File: fraudulent_activity_notifications/median_heap.c

#define TRUE (1)
#define FALSE (0)

typedef struct BinaryHeapNode_t BinaryHeapNode;
typedef struct BinaryHeap_t BinaryHeap;

struct BinaryHeapNode_t
{
    int value;
    int index;
    BinaryHeap* heap;
};

struct BinaryHeap_t
{
    int is_max_heap;
    int capacity;
    int count;
    BinaryHeapNode** heap;
};

typedef struct MovingMedian_t
{
    int max_samples;
    BinaryHeapNode** samples;
    int sample_index;
    BinaryHeap* max_heap;
    BinaryHeap* min_heap;
} MovingMedian;

// binary heap operations
static void init_heap(BinaryHeap* heap, int is_max_heap, int capacity);
static void destroy_heap(BinaryHeap* heap);
static BinaryHeapNode* add(BinaryHeap* heap, int value);
static void update(BinaryHeap* heap, BinaryHeapNode* node, int value);
static BinaryHeapNode* get_root(BinaryHeap* heap);
static void swap_root(BinaryHeap* heap, BinaryHeap* other_heap);
static void _update(BinaryHeap* heap, BinaryHeapNode* node, int should_move_up);
static BinaryHeapNode* _get_swap_up(BinaryHeap* heap, BinaryHeapNode* node);
static BinaryHeapNode* _get_swap_down(BinaryHeap* heap, BinaryHeapNode* node);
static int _should_move_up(BinaryHeap* heap, int first_value, int second_value);
static int _should_move_down(BinaryHeap* heap, int first_value, int second_value);
static void _swap_nodes(BinaryHeap* heap, BinaryHeapNode* node, BinaryHeapNode* other_node);
static BinaryHeapNode* _get_parent(BinaryHeap* heap, BinaryHeapNode* node);
static BinaryHeapNode* _get_left_child(BinaryHeap* heap, BinaryHeapNode* node);
static BinaryHeapNode* _get_right_child(BinaryHeap* heap, BinaryHeapNode* node);
static BinaryHeapNode* _get_max_or_min_child(BinaryHeap* heap, BinaryHeapNode* node);

// moving median operations
static void init_moving_median(MovingMedian* median, int max_samples);
static void destroy_moving_median(MovingMedian* median);
static void add_sample(MovingMedian* median, int value);
static int get_twice_median(MovingMedian* median);
static int get_sample_count(MovingMedian* median);

// Initialize a binary heap
static void init_heap(BinaryHeap* heap, int is_max_heap, int capacity)
{
    heap->is_max_heap = is_max_heap;
    heap->capacity = capacity;
    heap->count = 0;

    // caller must make sure capacity is greater than 0
    heap->heap = (BinaryHeapNode**)calloc(capacity, sizeof(BinaryHeapNode*));
}

static void destroy_heap(BinaryHeap* heap)
{
    // free memory allocated for each node
    for (int i = 0; i < heap->count; i++)
    {
        free(heap->heap[i]);
    }

    // free overall heap array
    free(heap->heap);
}

// adds an element to the heap
static BinaryHeapNode* add(BinaryHeap* heap, int value)
{
    BinaryHeapNode* result = 0;

    if (heap->count < heap->capacity)
    {
        // add the node as the first element in the heap array
        BinaryHeapNode* node = (BinaryHeapNode*)malloc(sizeof(BinaryHeapNode));
        node->value = value;
        node->index = heap->count;
        node->heap = heap;

        heap->heap[node->index] = node;
        heap->count += 1;

        // now fix up the heap to maintain the heap property
        _update(heap, node, TRUE);

        result = node;
    }

    return result;
}

// updates the value of a node in the heap
static void update(BinaryHeap* heap, BinaryHeapNode* node, int value)
{
    int oldValue = node->value;
    // dont need to do anything if value is the same
    if (oldValue != value)
    {
        node->value = value;
        _update(heap, node, _should_move_up(heap, value, oldValue));
    }
}

// get root node of heap or 0 if heap is empty
static BinaryHeapNode* get_root(BinaryHeap* heap)
{
    BinaryHeapNode* result = 0;

    if (heap->count > 0)
    {
        result = heap->heap[0];
    }

    return result;
}

// swaps the root node of the two heaps
static void swap_root(BinaryHeap* heap, BinaryHeap* other_heap)
{
    BinaryHeapNode* our_root = get_root(heap);
    BinaryHeapNode* their_root = get_root(other_heap);

    if ((our_root != 0) && (their_root != 0))
    {
        BinaryHeapNode* our_root_temp = our_root;
        our_root = their_root;
        their_root = our_root_temp;
        our_root->heap = heap;
        their_root->heap = other_heap;
        heap->heap[0] = our_root;
        other_heap->heap[0] = their_root;

        // now fix up the heaps
        _update(heap, our_root, FALSE);
        _update(other_heap, their_root, FALSE);
    }
}

// updates the heap by moving the node up or down until heap property is satisfied
static void _update(BinaryHeap* heap, BinaryHeapNode* node, int should_move_up)
{
    int continue_swapping = TRUE;

    while(continue_swapping)
    {
        BinaryHeapNode* swap_with = should_move_up
            ? _get_swap_up(heap, node)
            : _get_swap_down(heap, node);
        if (swap_with != 0)
        {
            _swap_nodes(heap, node, swap_with);
        }
        else
        {
            continue_swapping = FALSE;
        }
    }
}

static BinaryHeapNode* _get_swap_up(BinaryHeap* heap, BinaryHeapNode* node)
{
    BinaryHeapNode* result = 0;

    BinaryHeapNode* parent = _get_parent(heap, node);
    if (parent != 0)
    {
        int should_swap = _should_move_up(heap, node->value, parent->value);
        if (should_swap)
        {
            result = parent;
        }
    }

    return result;
}

static BinaryHeapNode* _get_swap_down(BinaryHeap* heap, BinaryHeapNode* node)
{
    BinaryHeapNode* result = 0;

    BinaryHeapNode* candidate_child = _get_max_or_min_child(heap, node);
    if (candidate_child != 0)
    {
        int should_swap = _should_move_down(heap, node->value, candidate_child->value);
        if (should_swap)
        {
            result = candidate_child;
        }
    }

    return result;
}

static int _should_move_up(BinaryHeap* heap, int first_value, int second_value)
{
    int result;

    if (heap->is_max_heap)
    {
        result = (first_value > second_value);
    }
    else
    {
        result = (first_value < second_value);
    }

    return result;
}

static int _should_move_down(BinaryHeap* heap, int first_value, int second_value)
{
    int result;

    if (heap->is_max_heap)
    {
        result = (first_value < second_value);
    }
    else
    {
        result = (first_value > second_value);
    }

    return result;
}

static void _swap_nodes(BinaryHeap* heap, BinaryHeapNode* node, BinaryHeapNode* other_node)
{
    int old_index = node->index;
    int old_other_index = other_node->index;

    heap->heap[old_index] = other_node;
    other_node->index = old_index;
    heap->heap[old_other_index] = node;
    node->index = old_other_index;
}

static BinaryHeapNode* _get_parent(BinaryHeap* heap, BinaryHeapNode* node)
{
    BinaryHeapNode* result = 0;

    // if node is the root then it has no parent
    if (node->index > 0)
    {
        // even indexed nodes are always righthand children.
        int is_even_node = ((node->index % 2) == 0);
        int offset = is_even_node ? 1 : 0;
        int parent_index = (node->index - offset - 1) / 2;

        result = heap->heap[parent_index];
    }
    
    return result;
}

static BinaryHeapNode* _get_left_child(BinaryHeap* heap, BinaryHeapNode* node)
{
    BinaryHeapNode* result = 0;

    // odd indexed nodes are always lefthand children.
    // NOTE: Could have overflow here if the heap is huge
    int child_index = (node->index * 2) + 1;
    if (child_index < heap->capacity)
    {
        result = heap->heap[child_index];
    }

    return result;
}

static BinaryHeapNode* _get_right_child(BinaryHeap* heap, BinaryHeapNode* node)
{
    BinaryHeapNode* result = 0;

    // even indexed nodes are always righthand children.
    // NOTE: Could have overflow here if the heap is huge
    int child_index = (node->index * 2) + 2;
    if (child_index < heap->capacity)
    {
        result = heap->heap[child_index];
    }

    return result;
}

static BinaryHeapNode* _get_max_or_min_child(BinaryHeap* heap, BinaryHeapNode* node)
{
    BinaryHeapNode* result = 0;
    BinaryHeapNode* left_child = _get_left_child(heap, node);
    BinaryHeapNode* right_child = _get_right_child(heap, node);

    if (left_child == 0)
    {
        // right child may still be null, but thats ok, returning null indicates no children
        result = right_child;
    }
    else if (right_child == 0)
    {
        // left child may still be null, but thats ok, returning null indicates no children
        result = left_child;
    }
    else
    {
        // use the largest child for max heap and smallest for min heap.
        int use_left_child = _should_move_up(heap, left_child->value, right_child->value);
        result = use_left_child ? left_child : right_child;
    }

    return result;
}

// initialize moving median calculator
static void init_moving_median(MovingMedian* median, int max_samples)
{
    median->max_samples = max_samples;
    median->samples = (BinaryHeapNode**)calloc(max_samples, sizeof(BinaryHeapNode*));
    median->sample_index = 0;

    // if the sample size is odd, ensure the max heap has a size that is one greater
    int min_heap_capacity = max_samples / 2;
    int max_heap_capacity = min_heap_capacity;
    int is_odd = (max_samples % 2) == 1;
    if (is_odd)
    {
        max_heap_capacity += 1;
    }

    median->max_heap = (BinaryHeap*)malloc(sizeof(BinaryHeap));
    median->min_heap = (BinaryHeap*)malloc(sizeof(BinaryHeap));
    init_heap(median->max_heap, TRUE, max_heap_capacity);
    init_heap(median->min_heap, FALSE, min_heap_capacity);
}

static void destroy_moving_median(MovingMedian* median)
{
    free(median->samples);
    destroy_heap(median->max_heap);
    destroy_heap(median->min_heap);
    free(median->max_heap);
    free(median->min_heap);
}

// adds a new sample to the moving median calculation. if the sample size has reached the capacity,
// the oldest sample will be replaced.
static void add_sample(MovingMedian* median, int value)
{
    int num_samples = get_sample_count(median);
    BinaryHeap* min_heap = median->min_heap;
    BinaryHeap* max_heap = median->max_heap;

    if  (num_samples == median->max_samples)
    {
        // we have reached the max number of samples. replace the value in the oldest sample
        BinaryHeapNode* node_to_update = median->samples[median->sample_index];
        update(node_to_update->heap, node_to_update, value);
    }
    else
    {
        /* 
            we havent reached the max sample size yet.
            alternate between adding to min and max heap to keep the heaps the same size or
            the max heap one size larger to ensure the median value(s) are always stored in
            the roots.
        */
        int is_even = (num_samples % 2) == 0;
        BinaryHeap* heap_to_add = is_even ? max_heap : min_heap;
        BinaryHeapNode* new_node = add(heap_to_add, value);
        median->samples[median->sample_index] = new_node;
    }

    /*
        at this point, the min and max heaps both satisfy their heap property indepedently,
        but, it's possible the heap we added/updated the value in wasnt the correct heap to
        ensure the two middle values are at the heap roots. So swap the roots then fix up the
        heaps again, if necessary
    */
    BinaryHeapNode* min_heap_root = get_root(min_heap);
    BinaryHeapNode* max_heap_root = get_root(max_heap);
    if ((min_heap_root != 0) && (max_heap_root != 0))
    {
        // min heap root value should be greater than max heap root value
        if ((min_heap_root->value) < (max_heap_root->value))
        {
            swap_root(min_heap, max_heap);
        }
    }

    median->sample_index += 1;
    median->sample_index %= median->max_samples;
}

// returns twice the current median value. returns 0 if there are no samples yet
static int get_twice_median(MovingMedian* median)
{
    int num_max_samples = median->max_heap->count;
    int num_min_samples = median->min_heap->count;
    int result = 0;

    /* 
        The logic has been designed to
        1) keep the heaps the same size if the number of samples is even
        2) make the max heap one element larger if the number of samples is odd.
        when the number of samples is odd, the median value is stored in the root of the max
        heap and when the number of samples is even, the median value is the average of the
        values stored in the two heap roots.
    */
    if ((num_max_samples > 0) || (num_min_samples > 0))
    {
        BinaryHeapNode* max_heap_root = get_root(median->max_heap);
        int low_median = max_heap_root->value;
        int high_median = low_median;
        
        if (num_max_samples == num_min_samples)
        {
            BinaryHeapNode* min_heap_root = get_root(median->min_heap);
            high_median = min_heap_root->value;
        }

        result = low_median + high_median;
    }

    return result;
}

// returns current number of samples in the moving median calculation
static int get_sample_count(MovingMedian* median)
{
    return median->min_heap->count + median->max_heap->count;
}

File: python/median_heap.py
"""
Solves: https://www.hackerrank.com/challenges/fraudulent-activity-notifications/problem?isFullScreen=true
"""

from dataclasses import dataclass
from typing import Generic, Optional, TypeVar

TValue = TypeVar("TValue", int, float)

@dataclass
class BinaryHeapNode(Generic[TValue]):
    """A node in a BinaryHeap"""

    value: TValue
    """The value stored in the node"""
    index: int
    """The index of the node in the underlying heap array"""
    heap: "BinaryHeap"
    """The heap this node belongs to"""


class BinaryHeap(Generic[TValue]):
    """A binary heap structure supporting min and max heaps"""

    def __init__(self, is_max_heap: bool, capacity: int):
        """Constructs an instance of a binary heap"""
        if capacity < 1:
            raise ValueError("capacity must be greater than 0")

        self._is_max_heap = is_max_heap
        self._capacity = capacity
        self._count = 0
        self._heap: list[Optional[BinaryHeapNode[TValue]]] = [None] * capacity

    def get_count(self) -> int:
        """Gets current number of items in the heap"""
        return self._count

    def get_root(self) -> Optional[BinaryHeapNode[TValue]]:
        """Gets the root node in the heap, or None if heap is empty"""
        try:
            return self._heap[0]
        except:
            return None

    def add(self, value: TValue) -> BinaryHeapNode[TValue]:
        """Adds a new element to the heap"""
        if self._count == self._capacity:
            raise RuntimeError("Heap is full!")

        # add the node as the first element in the heap array
        node = BinaryHeapNode(value, self._count, self)
        self._heap[node.index] = node
        self._count += 1

        # now fix up the heap to maintain the heap property
        self._update(node, True)

        return node

    def update(self, node: BinaryHeapNode[TValue], value: TValue):
        """Updates the value in a node and fixes up the heap to maintain the heap property"""
        oldValue = node.value

        # dont need to do anything if value is the same
        if value != oldValue:
            node.value = value
            self._update(node, self._should_move_up(value, oldValue))

    def swap_root(self, other_heap: "BinaryHeap[TValue]"):
        """Swaps the root of this heap with the root of the other heap, fixing up both heaps to
        maintain the heap property.
        """
        ourRoot = self.get_root()
        theirRoot = other_heap.get_root()
        if (ourRoot is None) or (theirRoot is None):
            raise RuntimeError("One of the heaps is empty! Can't swap roots")

        tempOurRoot = ourRoot
        ourRoot = theirRoot
        theirRoot = tempOurRoot
        ourRoot.heap = self
        theirRoot.heap = other_heap
        self._heap[0] = ourRoot
        other_heap._heap[0] = theirRoot

        # now fix up the heaps
        self._update(ourRoot, False)
        other_heap._update(theirRoot, False)

    def _update(self, node: BinaryHeapNode[TValue], move_up: bool):
        """Updates the heap to maintain the heap property when a node's value changes.
        
        :param node: The node that was updated
        :param move_up: Indicates which direction to move the node (up or down the tree)
        """
        continue_swapping = True
        while continue_swapping:
            swap_with = self._get_swap_up(node) if move_up else self._get_swap_down(node)
            if swap_with is not None:
                self._swap_nodes(node, swap_with)
            else:
                continue_swapping = False
    
    def _get_swap_up(self, node: BinaryHeapNode[TValue]) -> Optional[BinaryHeapNode[TValue]]:
        """Returns the parent node of the specified node if the node needs to be moved up to satisfy
        the heap property. Returns None if the node does not need to be moved up.
        """
        result: Optional[BinaryHeapNode[TValue]] = None

        parent = self._get_parent(node)
        if (parent is not None) and self._should_move_up(node.value, parent.value):
            result = parent

        return result

    def _get_swap_down(self, node: BinaryHeapNode[TValue]) -> Optional[BinaryHeapNode[TValue]]:
        """Returns a child node of the specified node if the node needs to be moved down to satisfy
        the heap property. Returns None if the node does not need to be moved down.
        """
        result: Optional[BinaryHeapNode[TValue]] = None

        candidate_child = self._get_max_or_min_child(node)
        if (candidate_child is not None) and self._should_move_down(node.value, candidate_child.value):
            result = candidate_child
        
        return result
    
    def _should_move_up(self, first_value: TValue, second_value: TValue) -> bool:
        """Indicates if the first value should be moved up in the heap, relative to the second value, to
        satisfy the heap property.
        """
        if self._is_max_heap:
            result = first_value > second_value
        else:
            result = first_value < second_value
        
        return result

    def _should_move_down(self, first_value: TValue, second_value: TValue) -> bool:
        """Indicates if the first value should be moved down in the heap, relative to the second value, to
        satisfy the heap property.
        """
        if self._is_max_heap:
            result = first_value < second_value
        else:
            result = first_value > second_value
        
        return result

    def _swap_nodes(self, node: BinaryHeapNode[TValue], other_node: BinaryHeapNode[TValue]):
        """Swaps two connected nodes in the heap"""
        old_index = node.index
        old_other_index = other_node.index

        self._heap[old_index] = other_node
        other_node.index = old_index
        self._heap[old_other_index] = node
        node.index = old_other_index

    def _get_parent(self, node: BinaryHeapNode[TValue]) -> Optional[BinaryHeapNode[TValue]]:
        """Returns the parent of a given node, or None if the node is the root node"""
        if node.index < 1:
            return None
        
        # even indexed nodes are always righthand children.
        offset = 1 if ((node.index % 2) == 0) else 0
        parent_index = (node.index - offset - 1) // 2

        return self._heap[parent_index]
    
    def _get_left_child(self, node: BinaryHeapNode[TValue]) -> Optional[BinaryHeapNode[TValue]]:
        """Returns the left child of a given node, or None if the node doesnt have a left child"""
        result: BinaryHeapNode[TValue] = None

        # odd indexed nodes are always lefthand children.
        child_index = (node.index * 2) + 1
        if child_index < self._capacity:
            result = self._heap[child_index]

        return result

    def _get_right_child(self, node: BinaryHeapNode[TValue]) -> Optional[BinaryHeapNode[TValue]]:
        """Returns the right child of a given node, or None if the node doesnt have a right child"""
        result: BinaryHeapNode[TValue] = None

        # even indexed nodes are always righthand children.
        child_index = (node.index * 2) + 2
        if child_index < self._capacity:
            result = self._heap[child_index]

        return result
    
    def _get_max_or_min_child(self, node: BinaryHeapNode[TValue]) -> Optional[BinaryHeapNode[TValue]]:
        """Gets the child with the greater value (if max heap) or smaller value (min heap). Returns
        None if the node does not have any children."""
        left_child = self._get_left_child(node)
        right_child = self._get_right_child(node)

        if left_child is None:
            # right child may still be null, but thats ok, returning null indicates no children
            result = right_child
        elif right_child is None:
            # left child may still be null, but thats ok, returning null indicates no children
            result = left_child
        else:
            # use the largest child for max heap and smallest for min heap
            result = left_child if self._should_move_up(left_child.value, right_child.value) else right_child

        return result

class MedianHeap(Generic[TValue]):
    """A data structure that combines equal sized (or within 1) max and min heaps to calculate a
    moving median over N samples with logirithmic time complexity.
    """

    def __init__(self, capacity: int):
        """Construct an instance of a moving median heap"""
        if capacity < 1:
            raise ValueError("capacity must be greater than 0")
        
        self._capacity = capacity
        self._rolling_values: list[Optional[BinaryHeapNode[TValue]]] = [None] * capacity
        self._rolling_index = 0

        # if the sample size is odd, ensure the max heap has a size that is one greater
        min_heap_capacity = self._capacity // 2
        max_heap_capacity = min_heap_capacity
        if (self._capacity % 2) == 1:
            max_heap_capacity += 1
    
        self._max_heap = BinaryHeap[TValue](True, max_heap_capacity)
        self._min_heap = BinaryHeap[TValue](False, min_heap_capacity)

    def get_count(self) -> int:
        """Returns current number of samples"""
        return self._max_heap.get_count() + self._min_heap.get_count()

    def get_twice_median(self):
        """Returns twice the median value of the samples"""
        if self.get_count() == 0:
            raise RuntimeError("No samples yet!")
        
        # The logic has been designed to
        # 1) keep the heaps the same size if the number of samples is even
        # 2) make the max heap one element larger if the number of samples is odd.
        # when the number of samples is odd, the median value is stored in the root of the max
        # heap and when the number of samples is even, the median value is the average of the
        # values stored in the two heap roots.
        low_median = self._max_heap.get_root().value
        high_median = self._max_heap.get_root().value
        if self._max_heap.get_count() == self._min_heap.get_count():
            high_median = self._min_heap.get_root().value
        
        return low_median + high_median

    def add(self, value: TValue):
        """Adds a new sample to the moving median calculation, removing the oldest sample if necessary"""
        if self.get_count() == self._capacity:
            # we have reached the max number of samples
            # replace the value in the oldest sample
            node_to_update = self._rolling_values[self._rolling_index]
            node_to_update.heap.update(node_to_update, value)
        else:
            # we havent reached the max sample size yet.
            # alternate between adding to min and max heap to keep the heaps the same size or
            # the max heap one size larger to ensure the median value(s) are always stored in
            # the roots.
            is_even = (self.get_count() % 2) == 0
            heap_to_add = self._max_heap if is_even else self._min_heap
            self._rolling_values[self._rolling_index] = heap_to_add.add(value)
        
        # at this point, the min and max heaps both satisfy their heap property indepedently,
        # but, it's possible the heap we added/updated the value in wasnt the correct heap to
        # ensure the two middle values are at the heap roots. So swap the roots then fix up the
        # heaps again, if necessary.
        min_heap_root = self._min_heap.get_root()
        max_heap_root = self._max_heap.get_root()
        if (min_heap_root is not None) and (max_heap_root is not None):
            # min heap root value should be greater than max heap root value
            if min_heap_root.value < max_heap_root.value:
                self._min_heap.swap_root(self._max_heap)

        self._rolling_index += 1
        self._rolling_index %= self._capacity
